### ISSUE TITLE
[VS Code] Question: How do you prevent the default completions from showing up as duplicates

### ISSUE THREAD

**Issue by felixhao28** [MicrosoftDocs/intellicode#85]
Related: https://github.com/Microsoft/vscode/issues/45864#issuecomment-483998252

I learned from VS Code side that there is [currently no way for an extension to suppress completions provided by other extensions from showing up](https://github.com/Microsoft/vscode/issues/45864#issuecomment-373394992). Yet somehow intellicode did it.

Notice (image below) there should be a "placeholder" provided by ms-python.python plugin but the intellicode completion replaces it. The completion marked by a diamond is provided by Kite, which is a third-party plugin.

![image](https://user-images.githubusercontent.com/2417391/56276609-665c5980-6135-11e9-9e78-703c9112140d.png)

I know this issue board might not a good place to raise questions like this, but I don't know where else to get help. My apologies if it annoys anyone.

**Comment by jkeech** [comment]
You are correct that extensions cannot modify or suppress completion list items from other extensions. IntelliCode is not able to do this either. IntelliCode takes a different approach — we directly extend the other language servers/extensions and modify the completion list before they send the initial results back to VS Code. So VS Code gets the full list with some starred items from the beginning, rather than IntelliCode providing its own set of extra recommendations that VS Code merges in with the default list.

**Comment by felixhao28** [comment]
Thank you!

> we directly extend the other language servers/extensions...

That sounds like exactly what I am looking for, but how does that work? Is there any related document or samples?

**Comment by jkeech** [comment]
The extension mechanism is different for each language server. I’m not sure if all are documented, so you might need to post issues on each language server that you are looking to extend. The TypeScript one looks like it is documented here though: https://github.com/Microsoft/TypeScript/wiki/Writing-a-Language-Service-Plugin

**Comment by felixhao28** [comment]
Thanks again. I am mostly interested in Python. And it seems I need to call an undocumented `python._loadLanguageServerExtension` command to load a python language extension .Net Core library.

![image](https://user-images.githubusercontent.com/2417391/56339261-066ebd00-61e0-11e9-8cc8-94e543669284.png)

The library needs to have a `LanguageServerExtensionProvider` class and a `LanguageServerExtension` class which implements `HandleCompletionAsync` method.

![image](https://user-images.githubusercontent.com/2417391/56339300-3d44d300-61e0-11e9-9047-ae61a85a38ad.png)

I got these information following the code in intellicode. I failed to find related resources online.
